donkeycar/parts/gps.py

```diff
@@ -0,0 +1,641 @@
+            import matplotlib.pyplot as plt
+            #define Matplotlib figure and axis
+            ax = plt.subplot(111, aspect='equal')
+            
+            # plot the collected readings
+            plt.scatter(self.x, self.y)
+            
+            # plot the centroid
+            plt.plot(self.x_stats.mean, self.y_stats.mean, marker="+", markeredgecolor="green", markerfacecolor="green")
+            
+            # plot the range
+            bounds = Rectangle(
+                (self.x_stats.min, self.y_stats.min), 
+                self.x_stats.max - self.x_stats.min, 
+                self.y_stats.max - self.y_stats.min,
+                alpha=0.5,
+                edgecolor='red',
+                fill=False,
+                visible=True)
+            ax.add_artist(bounds)
+
+            # plot the ellipsoid 
+            ellipse = Ellipse(xy=(self.x_stats.mean, self.y_stats.mean),
+                          width=self.width, height=self.height,
+                          angle=self.theta)
+            ellipse.set_alpha(0.25)
+            ellipse.set_facecolor('green')
+            ax.add_artist(ellipse)
+
+    def is_in_waypoint_range(waypoints, x, y):
+        i = 0
+        for waypoint in waypoints:
+            if waypoint.is_in_range(x, y):
+                return True, i
+            i += 1
+        return False, -1
+
+    def is_in_waypoint_std(waypoints, x, y, std):
+        i = 0
+        for waypoint in waypoints:
+            if waypoint.is_in_std(x, y, std):
+                return True, i
+            i += 1
+        return False, -1
+
+    def is_in_waypoint(waypoints, x, y):
+        i = 0
+        for waypoint in waypoints:
+            if waypoint.is_inside(x, y):
+                return True, i
+            i += 1
+        return False, -1
+
+
+    def plot(waypoints):
+        """
+        Draw the waypoint ellipsoid
+        """
+        from matplotlib.patches import Ellipse
+        import matplotlib.pyplot as plt
+        ax = plt.subplot(111, aspect='equal')
+        for waypoint in waypoints:
+            waypoint.plot()
+        plt.show()
+
+
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument("-s", "--serial", type=str, required=True, help="Serial port address, like '/dev/tty.usbmodem1411'")
+    parser.add_argument("-b", "--baudrate", type=int, default=9600, help="Serial port baud rate.")
+    parser.add_argument("-t", "--timeout", type=float, default=0.5, help="Serial port timeout in seconds.")
+    parser.add_argument("-sp", '--samples', type=int, default = 5, help = "Number of samples per waypoint.")
+    parser.add_argument("-wp", "--waypoints", type=int, default = 0, help = "Number of waypoints to collect; > 0 to collect waypoints, 0 to just log position")
+    parser.add_argument("-th", "--threaded", action='store_true', help = "run in threaded mode.")
+    parser.add_argument("-db", "--debug", action='store_true', help = "Enable extra logging")
```

Comment:
So the debug flag should only do a couple of things associated with the __main__ self test.  I changes most logs to info.  There are only a couple of places where the debug flag is used for things I only want to show up in the self test.

Reply:
Do we really need such a fine separation? I would suggest that if `--debug` is set in the test then you just set you logger level to DEBUG so all messages >= DEBUG are getting logged. And the same thing would happen in the vehicle loop, too. Setting DEBUG in the vehicle loop is only for hard-core debugging b/c it prints an awful lot of messages. I wouldn't see a problem that it would also print your debug statements here, they are universally useful in order to understand what is going on.

Reply:
I don't want all other debug messages when I use this; I did that originally and it makes this output impossible to use.  That is why I added the flag.